---
title: 重学 Go 语言：基础篇-35
---
<article id="topicContainer" class="column_content"><h2 class="topic_title"></h2><div><p><div class="toc">
<ul>
<li><ul>
<li><a href="#319">3-19 正确认识错误</a><ul>
<li><a href="#19">19 错误分类</a></li>
<li><a href="#19-1">19 错误处理</a></li>
<li><a href="#19-2">19 使用实例或类型判断错误类别</a></li>
<li><a href="#19-3">19 检查错误值</a></li>
<li><a href="#19-4">19 检查错误类型</a></li>
<li><a href="#19-5">19 常量错误值</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</p>
<h2 id="319">3-19 正确认识错误</h2>
<p>什么是error？和exception有什么区别？</p>
<p>错误(异常)是一种“值”，属于正常逻辑返回(exception,error)</p>
<ul>
<li>Stuck in 70's</li>
<li>Errors just values</li>
</ul>
<p>error是从C语言一脉相承，它完全抛弃了结构化异常上的一些技术进步。error代表的是一种状态、一种“值”。比如在登录时返回有正确、用户名不存在、密码错误、账户过期、账户禁用等各种状态，登录失败是很正常的状态，属于正常逻辑的一部分。显然返回值不是错误而是一种值，所以错误是正常状态值，表达的是另一种状态。用错误来称呼异常不太合理，这两个词都代表着负面含义，负面含义并不意味着就是错误。我们用error就是自定义一个exception，无非exception是简单变量error是复合变量。</p>
<h3 id="19">19 错误分类</h3>
<p>通常情况下我们把错误分成两类，一种是可恢复的，一种是不可恢复的。</p>
<p>不可恢复的是彻彻底底的错误，比如连接数据库服务器网络有问题或者数据库服务器宕机，这种情况下应用程序没有办法恢复，遇到这种不可恢复的错误通常的做法是打印输出错误堆栈后直接让应用程序直接崩溃，因为应用程序在这种情况下继续运行没有任何意义。最好方式让程序出错，并由监控程序对错误返回值进行处理，对管理员做出预警，管理员查看调用堆栈信息进行排查。</p>
<p>对于不可恢复的错误使用什么样的策略去处理呢？如果应用程序什么都不做一直等着错误消失，管理员不知道应用程序是正确还是错误，因为程序不死掉的话，管理员不知道应用程序是什么状态或者需要依次排查服务器。如果数据库服务器不是在同一个公司呢？或者监控程序只监控应用程序不监控数据库呢？所以应用程序是否抛出错误和依赖的资源是否被监控不是一回事，当遇到不可恢复错误的时候，最好的方式不是每隔一段时间去重试而是让程序死掉，通过输出的调用堆栈信息和错误返回的状态码给监控系统明确的指示，管理员接收指示排查原因，当服务器恢复时，管理员可选择一种触发机制让应用程序启动执行。程序重试管理员根本不知道，打印信息和程序死掉对于监控系统完全是两种状态，除非监控系统专门针对打印信息做出处理。</p>
<p>可恢复性错误，比如说有一个除法，这种错误可以恢复的，最简单的，拦截这个错误通知调用方换一个参数。</p>
<h3 id="19-1">19 错误处理</h3>
<p>基本处理规则。</p>
<ul>
<li>将error和正常返回值分离，明确告知会有错误发生。</li>
<li>与exception相比，error不会中断执行流程。</li>
<li>通过变量或类型匹配错误。</li>
<li>总是检查所返回错误，避免后续逻辑出错。</li>
<li>正常返回值和错误可能同时返回。(os.File.Read)</li>
</ul>
<p>当一个函数同时返回两个值的时候，并没有一种规范来保证两个值是互斥的，例如<code>io</code>包的<code>Read</code>函数。没有统一的标准的情况下，可以写注释，可以通过代码的冗余来保证它的结果一致。</p>
<h3 id="19-2">19 使用实例或类型判断错误类别</h3>
<p>Go语言不支持结构化异常时候，我们通常做法有几种，第一种使用变量，变量默认情况下要么实现error接口，要么实现自定义类型。判断返回值错误可以通过<code>err == errVar</code>判断某个具体的错误，<code>type(err) is IO</code>判断属于哪一类的错误。第二种使用常量，尽可能避免使用魔法数字。</p>
<p>使用结构化异常的优点可在不同函数中拦截，对异常的处理很灵活，在调用堆栈任何一个环节使用装饰器捕获异常。比如Web应用通常是client端发送request，web server对请求做handler，执行logic，可能有db最后返回结果。这样我们可以捕获错误信息包括异常，发送到log中，同时为了返回正常的状态可以重新抛出异常。使用返回值的话只能在调用的地方做判断。</p>
<p>错误是正常的返回值，是一种很合理的状态值，是一种很合理的流程控制。我们可以利用错误控制程序执行流程，写代码时完全可以把错误当作正常逻辑来使用，当作正常流控制来使用。</p>
<h3 id="19-3">19 检查错误值</h3>
<pre><code class="go language-go">var errTest = errors.New("test")

func test() error {
    return errTest
}
func main() {
    err := test()
    if err == errTest {
    }
}
</code></pre>
<blockquote>
  <p>检查变量，而非内容。</p>
</blockquote>
<p>对于错误值检查通常有几种方法，第一种把错误定义为变量，检查是不是指向同一个变量。</p>
<p>这种方式没必要单独建立类型，直接来检查。不要通过内容来匹配错误，内容是输出给调试人员看的辅助信息，不是用来匹配的。</p>
<pre><code class="go language-go">func test() error {
    return errors.New("test")
}
func main() {
    err := test()
    if err != nil &amp;&amp; err.Error() == "test" {
    }
}
</code></pre>
<p>有很多人直接<code>errors.New</code>，通过内容来判断错误。任何时候不要通过内容来匹配。</p>
<h3 id="19-4">19 检查错误类型</h3>
<pre><code class="go language-go">func main() {
    err := test()
    if _, ok := err.(ErrSome); ok {
    }
    switch err.(type) {
    case nil: // no error
    case ErrSome: // do something...
    default: // unknown error
    }
}
</code></pre>
<blockquote>
  <p>通过自定义类型匹配处理方式。</p>
</blockquote>
<p>通过检查错误类型，错误是哪一类错误。需要定义类型，指明是哪一类错误，这是结构化异常最常见做法，抛出哪一类错误。</p>
<p>检查错误有两种方式。第一种通过类型转换，把错误转化成某个具体的类型。第二种通过switch语句处理。
<code>nil</code>表示没有错误，是某一类错误进行处理，或者不知道哪种错误。</p>
<h3 id="19-5">19 常量错误值</h3>
<pre><code class="go language-go">type Error string

func (e Error) Error() string { return string(e) }

const ErrEOF = Error("EOF")

func test() error {
    return ErrEOF
}
func main() {
    err := test()
    println(err == Error("EOF"))
}
</code></pre>
<blockquote>
  <p>通过“字符串”实现常量值。</p>
</blockquote>
<p>有一种不推荐的写法，基于字符串来定义类型并实现<code>error</code>接口。使用常量定义错误，常量最终会被展开，也意味着从变量匹配变成内容匹配。匹配内容从设计上来说和设计理念不相符。</p>
<p>使用全局变量定义错误实际上非常危险的，可能被拦截修改。</p></div></article>